### PR TITLE
First crack at rank restriction

### PR DESCRIPTION
**Pull request by justinmuskopf**
Easiest to view `sv_restrict_propcore.lua` raw due to all of the deletions.

Like @brandonsturgeon said in #1, I understand that this will likely be scrapped, but I think we need to get this one in motion... So here's my first take. :)

### CODE REVIEW

File: lua/autorun/server/sv_restricted_propcore_functions.lua [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+-- Regular rank restricted functions
+REGULAR_RESTRICTED_FUNCTIONS = ADMIN_RESTRICTED_FUNCTIONS
+REGULAR_RESTRICTED_FUNCTIONS['propSpawn(sn)']           = {build = true, pvp = true}
+REGULAR_RESTRICTED_FUNCTIONS['propSpawn(en)']           = {build = true, pvp = true}
+REGULAR_RESTRICTED_FUNCTIONS['propSpawn(svn)']          = {build = true, pvp = true}
+REGULAR_RESTRICTED_FUNCTIONS['propSpawn(evn)']          = {build = true, pvp = true}
+REGULAR_RESTRICTED_FUNCTIONS['propSpawn(san)']          = {build = true, pvp = true}
+REGULAR_RESTRICTED_FUNCTIONS['propSpawn(ean)']          = {build = true, pvp = true}
+REGULAR_RESTRICTED_FUNCTIONS['propSpawn(svan)']         = {build = true, pvp = true}
+REGULAR_RESTRICTED_FUNCTIONS['propSpawn(evan)']         = {build = true, pvp = true}
```
Comment (justinmuskopf) [inline lua/autorun/server/sv_restricted_propcore_functions.lua:33]: I have an idea for this.

Keep the full function signature (including parameters) in `RESTRICTED_FUNCTIONS`, but do something like this for the insertion into rank tables:
```lua
REGULAR_RESTRICTED_FUNCTIONS['propSpawn'] = {build = true, pvp = true}
```
And when checking the restrictions in the code, `gsub` the (parameters) from the function's signature, and check it then.
Unless you think there will be cases where we'll give someone (for example) `propSpawn(sn)` but not `propSpawn(en)`?

**Comment by sarahsturgeon** [comment]
@justinmuskopf I like this structure so far, but I'm worried about its modularity.

How easy is it to add another rank in here?

Also, could you try to resolve the conflicts with master?

File: lua/autorun/server/sv_restricted_propcore_functions.lua [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+REGULAR_RESTRICTED_FUNCTIONS['seatSpawn(sn)']           = {build = true, pvp = true}
+REGULAR_RESTRICTED_FUNCTIONS['seatSpawn(svan)']         = {build = true, pvp = true}
+REGULAR_RESTRICTED_FUNCTIONS['setPos(e:v)']             = {build = true, pvp = true}
+REGULAR_RESTRICTED_FUNCTIONS['propManipulate(e:vannn)'] = {build = true, pvp = true}
+REGULAR_RESTRICTED_FUNCTIONS['reposition(e:v)']         = {build = true, pvp = true}
+--REGULAR_RESTRICTED_FUNCTIONS['propBreak(e:)']         = {build = true, pvp = true}
+
+-- User rank restricted functions
+local USER_RESTRICTED_FUNCTIONS = REGULAR_RESTRICTED_FUNCTIONS
```
Comment (justinmuskopf) [inline lua/autorun/server/sv_restricted_propcore_functions.lua]: @brandonsturgeon 

Adding a new rank would be as so:
```lua
local NEWRANK_RESTRICTED_FUNCTIONS = RANK_THAT_SUPERSEDES_THIS_RESTRICTED_FUNCTIONS
NEWRANK_RESTRICTED_FUNCTIONS[function] = {build = t/f, pvp = t/f}
...

local RESTRICTED_FUNTIONS_BY_RANK = {
    admin = ...
    regular = ...
    user = ...
    newRank = NEWRANK_RESTRICTED_FUNCTIONS
}
```
The good thing about this is that you could base a rank off of  the `user` rank, and then override/add any methods for specific allowances.

Comment (sarahsturgeon) [reply lua/autorun/server/sv_restricted_propcore_functions.lua]: Nice, so it's based on an inheritance system.

That makes sense as long as we travel in one direction, but what if we want to allow something to Regulars but not Devotees? How challenging would that be?

Comment (justinmuskopf) [reply lua/autorun/server/sv_restricted_propcore_functions.lua]: Assuming that Regulars inherit from Devotees, they would initially have the same restriction table.
However, to allow something to them that isn't allowed to Devotees:

Where Devotees would have:
```lua
DEVOTEE_RESTRICTED_FUNCTIONS["function"] = {build = true, pvp = true}
```

The Regular setup might look like:
```
REGULAR_RESTRICTED_FUNCTIONS = DEVOTEE_RESTRICTED_FUNCTIONS

...

REGULAR_RESTRICTED_FUNCTIONS["function"] = {build = false, pvp = false}
```
The entry in their table would just be overwritten.

Example: https://imgur.com/drEAa1m

Comment (justinmuskopf) [reply lua/autorun/server/sv_restricted_propcore_functions.lua]: Just found out these boys are passed by reference. We'll need a table copy function.

Comment (justinmuskopf) [reply lua/autorun/server/sv_restricted_propcore_functions.lua]: Okay, that's done.

File: lua/autorun/server/sv_restricted_propcore_functions.lua [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+local function copyTableByValue(tab)
+    returnTable = {}
+
+    -- Copy by value
+    for key, value in pairs( tab ) do
+        returnTable[key] = value
+    end
+
+    return returnTable
+end
```
Comment (justinmuskopf) [inline lua/autorun/server/sv_restricted_propcore_functions.lua]: I've confirmed that this works.
The tables, when assigned by `=`, were being assigned by reference, not value.

Not anymore!

**Review by sarahsturgeon — APPROVED** [review]
Approving. Seems like you already tested it?

**Comment by brandonsturgeon** [comment]
resolv confic

**Comment by justinmuskopf** [comment]
there are no confic

**Comment by brandonsturgeon** [comment]
image

**Comment by sarahsturgeon** [comment]
@justinmuskopf FIx conflic

@OverlordBleck @PLally Can you two test this branch on CFC3?
If so, let me know and I can swap the server version over to this branch for testing.

File: lua/autorun/server/sv_restrict_propcore.lua [hunk 4]
```diff
@@ -1,41 +1,40 @@
...
+    for _, signature in pairs( RESTRICTED_FUNCTIONS ) do
         local oldFunc = wire_expression2_funcs[signature][3]
-
         wire_expression2_funcs[signature][3] = function( self, ... )
+            local playerGroup       = self.player:GetUserGroup()
+            local groupRestrictions = rankRestrictedFunctions[playerGroup] or rankRestrictedFunctions['user']
+            local restrictions      = groupRestrictions[signature]
+
             if ( disallowedRanks[self.player:GetUserGroup()] == nil ) then
                 local isInBuildMode = self.player:GetNWBool("CFC_PvP_Mode", false) == false
```
Comment (plally) [inline lua/autorun/server/sv_restrict_propcore.lua]: You forgot an `end`

File: lua/autorun/server/sv_restrict_propcore.lua [hunk 5]
```diff
@@ -1,41 +1,40 @@
...
-        "seatSpawn(sn)",
-        "seatSpawn(svan)",
-        "setPos(e:v)",
-        "reposition(e:v)",
-        "propManipulate(e:vannn)"
-        --"propBreak(e:)"
-    }
-
-    for _, signature in pairs( restrictedFunctions ) do
+    for _, signature in pairs( RESTRICTED_FUNCTIONS ) do
```
Comment (plally) [inline lua/autorun/server/sv_restrict_propcore.lua]:
```suggestion
    for _, signature in pairs( restrictedFunctions ) do
```

File: lua/autorun/server/sv_restrict_propcore.lua [hunk 6]
```diff
@@ -1,41 +1,40 @@
...
         local oldFunc = wire_expression2_funcs[signature][3]
-
         wire_expression2_funcs[signature][3] = function( self, ... )
+            local playerGroup       = self.player:GetUserGroup()
+            local groupRestrictions = rankRestrictedFunctions[playerGroup] or rankRestrictedFunctions['user']
+            local restrictions      = groupRestrictions[signature]
+
             if ( disallowedRanks[self.player:GetUserGroup()] == nil ) then
```
Comment (plally) [inline lua/autorun/server/sv_restrict_propcore.lua]: disallowedRanks ranks isn't defined  and this statement seems to have no affect as isInBuildMode is defined later in the function

**Review by plally — CHANGES_REQUESTED** [review]

File: lua/autorun/server/sv_restrict_propcore.lua [hunk 7]
```diff
@@ -1,41 +1,40 @@
--- Propcore is allowed to everyone, but functions in the restrictedFunctions array will be restricted to devotee+ only
+-- Restricts PropCore functions based on rank
 
+local restrictedFunctions, rankRestrictedFunctions = require "sv_restricted_propcore_functions"
```
Comment (plally) [inline lua/autorun/server/sv_restrict_propcore.lua]: this require wasn't working for me`[ERROR] addons/cfc_restrict_propcore/lua/autorun/server/sv_restrict_propcore.lua:3: Module not found!`
but this worked
```lua
local restrictedFunctions, rankRestrictedFunctions = include "sv_restricted_propcore_functions.lua"
```

**Comment by sarahsturgeon** [comment]
@justinmuskopf can you address hmm's review?

File: lua/autorun/server/sv_restrict_propcore.lua (hunk 7 above)
Comment (justinmuskopf) [reply lua/autorun/server/sv_restrict_propcore.lua]: /shrug

File: lua/autorun/server/sv_restrict_propcore.lua (hunk 4 above)
Comment (justinmuskopf) [reply lua/autorun/server/sv_restrict_propcore.lua]: You forgot my birthday.

**Comment by justinmuskopf** [comment]
> @justinmuskopf can you address hmm's review?

There. Happy?

**Comment by sarahsturgeon** [comment]
_could you also fix the conflicts_

**Comment by justinmuskopf** [comment]
done hands

**Comment by plally** [comment]
retest this please

**Comment by sarahsturgeon** [comment]
It appears this ol' dinosaur has to be put down. Rank restrictions were added in another PR.

**Pull request closed without merging**